### DISCUSSION TITLE
Manually invokable jobs

### DISCUSSION THREAD

**Discussion by ericallam** [triggerdotdev/trigger.dev#516] [Ideas]
Currently there are 3 ways of triggering a job run:

- Events
- On a schedule
- Externally (webhooks)

We've had multiple requests for a way to manually invoke a job, passing in a payload, and getting back a single run. This would work similarly to how our job test service works, but work via API and within another run. 

## Invoke Trigger

Jobs that can be manually invoked will use a new Trigger called `invokeTrigger`. This will work similarly to `eventTrigger` but won't have an event name:

```ts
import { TriggerClient, invokeTrigger } from "@trigger.dev/sdk";
import { z } from "zod";

export const client = new TriggerClient({
  id: "job-catalog",
});

export const invokeExample1 = client.defineJob({
  id: "invoke-example-1",
  name: "Invoke Example 1",
  version: "1.0.0",
  enabled: true,
  trigger: invokeTrigger({
    schema: z.object({
      name: z.string(),
    }),
  }),
  run: async (payload, io, ctx) => {
    console.log(payload); // { name: "John" }
  },
});
```

Now that job can be manually invoked using a reference to the `Job` instance:

```ts
import { invokeExample1 } from "./invoke";

// This would return the run immediately, and can be used outside or inside of another job run
const run = await invokeExample1.invoke({ name: "John" })

// This would return the run after it's completed, and can only be used inside of another job run
// If the invokeExample1 fails, the invocation task would fail. If it completes successfully, this would return the job output (typed correctly)
const output = await invokeExample1.invokeAndWait("invoke-job", { name: "John" })
```

Another benefit of having this manually invokable jobs is it would make it much more ergonomic to use with the [useRunDetails](https://trigger.dev/docs/documentation/guides/react-hooks-automatic#userundetails) React hook.

## Implementation

This could work similarly to the `TestJobService`, where an event is created and then immediately used to create a new run, instead of going through the `EventDispatcher` system. The name of the event could be the job id, and the `id` of the event would either be randomly generated on the server, or the invocation task idempotency key when invoked from inside another job run.

**Comment by matt-aitken** [comment]
This was implemented 🥳
